yjit/src/codegen.rs

```diff
@@ -2799,7 +2799,7 @@ fn gen_defined(
         _ => {
             // Save the PC and SP because the callee may allocate
             // Note that this modifies REG_SP, which is why we do it first
-            jit_prepare_non_leaf_call(jit, asm);
+            jit_prepare_call_with_gc(jit, asm);
```

Comment:
Hopefully this is issue causing assert_ctx to trip?

`verify_ctx: ctx type (Fixnum) incompatible with actual value of local: T_NIL (type Nil)`

You can call methods with side effects inside `defined?`.

Reply:
Yeah, that was it https://github.com/ruby/ruby/pull/9874/commits/3567148a38d7efd96e4e58af1f10ca959fc13031. 

I knew expressions in `defined?` can make a call, but ISEQs have method calls before `defined` instruction, so I still thought `defined` instruction itself is leaf. It turns out it could still make a `respond_to?` call inside `defined` instructions, so it was not leaf.

Maybe we could still switch this if `defined_type` is not `DEFINED_FUNC` or `DEFINED_METHOD`, but that's too specific to `defined` instruction, so I'll leave it to another PR.